File: v1/LLD/Design-Patterns/Bridge/code/Bridge.java
// Bridge Pattern — Shape abstraction bridged to a Renderer implementor.
// Compile: javac Bridge.java
// Run:     java Bridge

// Implementor hierarchy
interface Renderer {
    void renderCircle(float radius);
    void renderSquare(float side);
}

class OpenGLRenderer implements Renderer {
    public void renderCircle(float radius) {
        System.out.println("[OpenGL] drawing circle, radius=" + radius);
    }
    public void renderSquare(float side) {
        System.out.println("[OpenGL] drawing square, side=" + side);
    }
}

class DirectXRenderer implements Renderer {
    public void renderCircle(float radius) {
        System.out.println("[DirectX] drawing circle, radius=" + radius);
    }
    public void renderSquare(float side) {
        System.out.println("[DirectX] drawing square, side=" + side);
    }
}

// Abstraction hierarchy
abstract class Shape {
    protected Renderer renderer;

    Shape(Renderer renderer) {
        this.renderer = renderer;
    }

    abstract void draw();
}

class Circle extends Shape {
    private float radius;

    Circle(Renderer renderer, float radius) {
        super(renderer);
        this.radius = radius;
    }

    void draw() {
        renderer.renderCircle(radius);
    }
}

class Square extends Shape {
    private float side;

    Square(Renderer renderer, float side) {
        super(renderer);
        this.side = side;
    }

    void draw() {
        renderer.renderSquare(side);
    }
}

public class Bridge {
    public static void main(String[] args) {
        Shape openGLCircle = new Circle(new OpenGLRenderer(), 5.0f);
        Shape directXSquare = new Square(new DirectXRenderer(), 3.0f);

        openGLCircle.draw();
        directXSquare.draw();
    }
}

File: v1/LLD/Design-Patterns/Bridge/code/bridge.js
/**
 * Bridge Pattern — Shape abstraction bridged to a Renderer implementor.
 * Run: node bridge.js
 */

class OpenGLRenderer {
  renderCircle(radius) {
    console.log(`[OpenGL] drawing circle, radius=${radius}`);
  }
  renderSquare(side) {
    console.log(`[OpenGL] drawing square, side=${side}`);
  }
}

class DirectXRenderer {
  renderCircle(radius) {
    console.log(`[DirectX] drawing circle, radius=${radius}`);
  }
  renderSquare(side) {
    console.log(`[DirectX] drawing square, side=${side}`);
  }
}

class Shape {
  constructor(renderer) {
    this.renderer = renderer;
  }
}

class Circle extends Shape {
  constructor(renderer, radius) {
    super(renderer);
    this.radius = radius;
  }
  draw() {
    this.renderer.renderCircle(this.radius);
  }
}

class Square extends Shape {
  constructor(renderer, side) {
    super(renderer);
    this.side = side;
  }
  draw() {
    this.renderer.renderSquare(this.side);
  }
}

const openGLCircle = new Circle(new OpenGLRenderer(), 5.0);
const directXSquare = new Square(new DirectXRenderer(), 3.0);

openGLCircle.draw();
directXSquare.draw();

module.exports = { Circle, Square, OpenGLRenderer, DirectXRenderer };
